kernel/kernel-api/src/main/java/io/delta/kernel/internal/checksum/ChecksumUtils.java

```diff
@@ -0,0 +1,128 @@
+    LongAdder tableSizeByte = new LongAdder();
+    LongAdder fileCount = new LongAdder();
+    FileSizeHistogram fileSizeHistogram = FileSizeHistogram.createDefaultHistogram();
+    Map<String, DomainMetadata> domainMetadataMap = new HashMap<>();
+    ChecksumWriter checksumWriter = new ChecksumWriter(snapshot.getLogPath());
+
+    // snapshot.getLogSegment() could return last available CRC.
+    // Set minFileRetentionTimestampMillis to infinite future to skip all removed files.
+    try (CreateCheckpointIterator checkpointIterator =
+        new CreateCheckpointIterator(
+            engine,
+            snapshot.getLogSegment(),
+            Instant.ofEpochMilli(Long.MAX_VALUE).toEpochMilli())) {
+      checkpointIterator.forEachRemaining(
+          batch ->
+              batch
+                  .getRows()
+                  .forEachRemaining(
+                      row -> {
+                        checkState(
+                            row.isNullAt(CHECKPOINT_SCHEMA.indexOf("remove")),
+                            "unexpected remove row found when setting "
+                                + "minFileRetentionTimestampMillis to infinite future");
+                        if (!row.isNullAt(CHECKPOINT_SCHEMA.indexOf("add"))) {
+                          long addFileSize =
+                              new AddFile(row.getStruct(CHECKPOINT_SCHEMA.indexOf("add")))
+                                  .getSize();
+                          tableSizeByte.add(addFileSize);
+                          fileSizeHistogram.insert(addFileSize);
+                          fileCount.increment();
+                        }
+                        if (!row.isNullAt(CHECKPOINT_SCHEMA.indexOf("domainMetadata"))) {
+                          DomainMetadata domainMetadata =
+                              DomainMetadata.fromRow(
+                                  row.getStruct(CHECKPOINT_SCHEMA.indexOf("domainMetadata")));
+                          if (!domainMetadataMap.containsKey(domainMetadata.getDomain())) {
+                            domainMetadataMap.put(domainMetadata.getDomain(), domainMetadata);
+                          }
+                        }
+                      }));
+      checksumWriter.writeCheckSum(
+          engine,
+          new CRCInfo(
+              snapshot.getVersion(),
+              snapshot.getMetadata(),
+              snapshot.getProtocol(),
+              tableSizeByte.longValue(),
+              fileCount.longValue(),
+              Optional.empty() /* txnId */,
+              Optional.of(
+                  domainMetadataMap.values().stream()
+                      .filter(domainMetadata -> !domainMetadata.isRemoved())
+                      .collect(Collectors.toSet())),
+              Optional.of(fileSizeHistogram)));
+
+    } catch (FileAlreadyExistsException fileAlreadyExistsException) {
+      throw new ChecksumAlreadyExistsException(snapshot.getVersion());
+    }
+  }
+
+  private ChecksumUtils() {}
```

Comment:
For a utility class with only public static helper methods, let's put the private constructor at the top. I'll take note of this and add this to a dev guide / java style guide I am working on.